autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/service/AutofillProviderDeviceAuth.kt

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (c) 2025 DuckDuckGo
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.autofill.impl.service
+
+import androidx.datastore.core.DataStore
+import androidx.datastore.preferences.core.Preferences
+import androidx.datastore.preferences.core.edit
+import androidx.datastore.preferences.core.longPreferencesKey
+import com.duckduckgo.autofill.impl.deviceauth.AutofillDeviceAuthStore
+import com.duckduckgo.autofill.impl.time.TimeProvider
+import com.duckduckgo.di.scopes.AppScope
+import com.squareup.anvil.annotations.ContributesBinding
+import dagger.SingleInstanceIn
+import javax.inject.Inject
+import kotlinx.coroutines.flow.firstOrNull
+import timber.log.Timber
+
+interface AutofillProviderDeviceAuth {
+    suspend fun isAuthRequired(): Boolean
+    suspend fun recordSuccessfulAuthorization()
+}
+
+@SingleInstanceIn(AppScope::class)
+@ContributesBinding(AppScope::class)
+class RealAutofillProviderDeviceAuth @Inject constructor(
+    private val timeProvider: TimeProvider,
+    @AutofillDeviceAuthStore private val store: DataStore<Preferences>,
+) : AutofillProviderDeviceAuth {
+
+    private val lastAuthTimeKey = longPreferencesKey("last_auth_time_key")
+
+    override suspend fun isAuthRequired(): Boolean {
+        val lastAuthTime = store.data.firstOrNull()?.get(lastAuthTimeKey) ?: 0L
+        if (lastAuthTime == 0L) return true
+
+        val timeSinceLastAuth = timeProvider.currentTimeMillis() - lastAuthTime
+        if (timeSinceLastAuth <= AUTH_GRACE_EXTENDED_PERIOD_MS) {
+            Timber.v("Within grace period; auth not required")
+            return false
+        }
+        return true
+    }
+
+    override suspend fun recordSuccessfulAuthorization() {
+        store.edit {
+            it[lastAuthTimeKey] = timeProvider.currentTimeMillis()
+        }
+    }
+
+    companion object {
+        private const val AUTH_GRACE_EXTENDED_PERIOD_MS = 180_000
```

Comment:
how did we land on 180s for this? not saying it’s bad (i’m not sure what my intuition would be if i was picking it) but curious why this value…was it similar to other providers for instance?

Reply:
I don't know if there's a standard here. Some apps allow users to decide, others will go for a default value but I haven't checked.

I just landed in the same value we use for import passwords from google as a way to avoid asking on every field in the same app. I will leave it documented in https://app.asana.com/0/1200156640058969/1209180548400497, to indicate there's no reasoning further than that, and we can change if we are not happy with it.